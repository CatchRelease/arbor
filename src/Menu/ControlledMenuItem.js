import React from 'react';
import PropTypes from 'prop-types';
import { hideVisually } from 'polished';

import { ENTER_KEY, SPACEBAR } from '../constants';
import Flex from '../Flex';
import Icon from '../Icon';
import MenuItem from '../MenuItem';
import Text from '../Text';

const MIN_WIDTH = '120px';
const PADDING = '8px';

class ControlledMenuItem extends React.PureComponent {
  componentDidMount() {
    document.addEventListener('keydown', this.onKeyDown, true);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.onKeyDown, true);
  }

  onKeyDown = (e) => {
    const { key } = e;
    const { isFocused } = this.props;

    if (!isFocused) {
      return;
    }

    if (key === ENTER_KEY || key === SPACEBAR) {
      e.preventDefault();
      this.onSelect();
    }
  };

  onSelect = () => {
    const { onSelect, value } = this.props;
    onSelect(value);
  };

  render() {
    const {
      MenuItemComponent,
      baseColor,
      iconName,
      id,
      isFocused,
      isSelected,
      label,
      name,
      secondaryLabel,
      value
    } = this.props;

    return (
      <MenuItemComponent
        {...{
          'aria-selected': isSelected,
          alignItems: 'center',
          isFocused,
          isSelected,
          onClick: this.onSelect,
          onKeyPress: this.onKeyDown,
          role: 'option',
          ...this.props
        }}
      >
        <input {...{ id, name, value, css: hideVisually(), type: 'hidden' }} />
        <Flex width="100%" justifyContent="space-between">
          <Flex
            flex="1 0 auto"
            maxWidth={
              secondaryLabel ? `calc(100% - ${MIN_WIDTH} - ${PADDING})` : '100%'
            }
          >
            {iconName && (
              <Icon
                name={iconName}
                mr="small"
                color={baseColor}
                fontSize="16px"
              />
            )}
            <Text
              as="span"
              color="inherit"
              minWidth={MIN_WIDTH}
              fontSize="size4"
              overflow="hidden"
            >
              {label}
            </Text>
          </Flex>
          {secondaryLabel && (
            <Flex pl={PADDING} overflow="hidden" flex="0 1 auto">
              <Text
                as="span"
                fontSize="size3"
                css={(theme) => ({
                  // Put primary and secondary labels on the same baseline
                  lineHeight: theme.fontSizes.size4
                })}
                color="text.muted"
                maxWidth="100%"
              >
                {secondaryLabel}
              </Text>
            </Flex>
          )}
        </Flex>
      </MenuItemComponent>
    );
  }
}

ControlledMenuItem.propTypes = {
  /**
   * Base color to use for the menu item
   */
  baseColor: PropTypes.string,

  /**
   * Name of icon to use for the menu item
   */
  iconName: PropTypes.string,

  /**
   * HTML ID for the menu item
   */
  id: PropTypes.string.isRequired,

  /**
   * Whether or not the current menu item is focused
   */
  isFocused: PropTypes.bool,

  /**
   * Indicates whether or not the menu item is selected
   */
  isSelected: PropTypes.bool,

  /**
   * Main label text for the menu item
   */
  label: PropTypes.string.isRequired,

  /**
   * Component used to render each menu item
   */
  MenuItemComponent: PropTypes.elementType,

  /**
   * HTML input name property for the input field
   */
  name: PropTypes.string.isRequired,

  /**
   * Callback to run when a selection is made. This is generally intended for
   * internal use with the Menu component.
   */
  onSelect: PropTypes.func,

  /**
   * Secondary label text for the menu item
   */
  secondaryLabel: PropTypes.string,

  /**
   * HTML input value property for the input field
   */
  value: PropTypes.string.isRequired
};

ControlledMenuItem.defaultProps = {
  MenuItemComponent: MenuItem,
  baseColor: 'text.default',
  iconName: null,
  isFocused: false,
  isSelected: false,
  onSelect: () => null,
  secondaryLabel: null
};

export default ControlledMenuItem;
