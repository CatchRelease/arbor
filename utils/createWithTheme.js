import React from 'react';
import { CacheProvider } from '@emotion/core';
import createCache from '@emotion/cache';
import { create } from 'react-test-renderer';
import { ThemeProvider } from 'emotion-theming';
import theme from '../src/theme';

const myCache = createCache();
myCache.compat = true;

export default (Component) =>
  create(
    <CacheProvider value={myCache}>
      <ThemeProvider theme={theme}>{Component}</ThemeProvider>
    </CacheProvider>
  );
