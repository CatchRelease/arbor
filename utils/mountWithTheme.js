import { mount } from 'enzyme';
import { ThemeProvider } from '@emotion/react';
import theme from '../src/theme';

export default (Component, mountOptions = {}) => {
  const wrappedComponent = (
    <ThemeProvider theme={theme}>{Component}</ThemeProvider>
  );

  const mounted = mount(wrappedComponent, mountOptions);

  return mounted.find(Component.type);
};
