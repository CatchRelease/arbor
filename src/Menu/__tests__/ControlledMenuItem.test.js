import React from 'react';

import mountWithTheme from '../../../utils/mountWithTheme';
import createWithTheme from '../../../utils/createWithTheme';
import { ENTER_KEY, SPACEBAR } from '../../constants';
import ControlledMenuItem from '../ControlledMenuItem';

describe('<ControlledMenuItem />', () => {
  const noop = () => null;
  const baseProps = {
    label: 'Ready to Clear',
    secondaryLabel: 'mpd@catchandrelease.com',
    baseColor: 'grey',
    name: 'clearance_state',
    value: 'ready_to_clear',
    id: 'menu-item_clearance_state_ready_to_clear'
  };

  function createWithProps(additionalProps) {
    const props = Object.assign({}, baseProps, additionalProps);
    return createWithTheme(<ControlledMenuItem {...{ ...props }} />);
  }

  function mountWithProps(additionalProps) {
    const props = Object.assign({}, baseProps, additionalProps);
    return mountWithTheme(<ControlledMenuItem {...{ ...props }} />);
  }

  it('renders a MenuItem properly', () => {
    const menuItem = createWithProps();

    expect(menuItem).toMatchSnapshot();
  });

  it('renders a selected MenuItem properly', () => {
    const menuItem = createWithProps({ selected: true });

    expect(menuItem).toMatchSnapshot();
  });

  describe('onClick', () => {
    it('calls the onSelect callback with the value', () => {
      const onSelect = jest.fn();
      const value = 'ready_to_clear';
      const menuItem = mountWithProps({ onSelect, value });

      menuItem.simulate('click');

      expect(onSelect).toHaveBeenCalledWith(value);
    });
  });

  describe('keydown', () => {
    const value = 'ready_to_clear';

    const behavesLikeKeyboardSelector = key => {
      context('menu item is focused', () => {
        it('calls the onSelect callback', () => {
          const onSelect = jest.fn();
          const menu = mountWithProps({ onSelect, value, isFocused: true });

          menu.instance().onKeyDown({ key, preventDefault: noop });

          expect(onSelect).toHaveBeenCalledWith(value);
        });

        it('prevents the default keydown event', () => {
          const menu = mountWithProps({ isFocused: true });
          const preventDefault = jest.fn();

          menu.instance().onKeyDown({ key, preventDefault });

          expect(preventDefault).toHaveBeenCalled();
        });
      });

      context('menu item is not focused', () => {
        it('does not call the onSelect callback', () => {
          const onSelect = jest.fn();
          const menu = mountWithProps({ onSelect, value, isFocused: false });

          menu.instance().onKeyDown({ key });

          expect(onSelect).not.toHaveBeenCalled();
        });
      });
    };

    context('Enter key', () => {
      behavesLikeKeyboardSelector(ENTER_KEY);
    });

    context('Spacebar', () => {
      behavesLikeKeyboardSelector(SPACEBAR);
    });

    context('Other key than Enter or space', () => {
      context('menu item is focused', () => {
        it('does not call the onSelect callback', () => {
          const onSelect = jest.fn();
          const menu = mountWithProps({ onSelect, value, isFocused: true });

          menu.instance().onKeyDown({ key: 'ArrowDown' });

          expect(onSelect).not.toHaveBeenCalled();
        });
      });
    });
  });
});
