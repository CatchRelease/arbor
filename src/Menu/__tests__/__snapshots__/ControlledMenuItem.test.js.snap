// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ControlledMenuItem /> renders a MenuItem properly 1`] = `
.emotion-11 {
  font-family: Motiva Sans;
  font-size: 0.875rem;
  padding: 0.75rem;
  box-sizing: border-box;
  background: #FFFFFF;
  color: #25272D;
  cursor: pointer;
}

.emotion-11:hover,
.emotion-11:focus {
  background: #F8F9F9;
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-9 {
  font-family: Motiva Sans;
  box-sizing: border-box;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.emotion-3 {
  font-family: Motiva Sans;
  max-width: calc(100% - 120px - 8px);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

.emotion-1 {
  font-family: Motiva Sans;
  color: inherit;
  font-size: 0.875rem;
  line-height: 1.25;
  min-width: 120px;
  overflow: hidden;
}

.emotion-7 {
  font-family: Motiva Sans;
  overflow: hidden;
  padding-left: 8px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
}

.emotion-5 {
  font-family: Motiva Sans;
  color: #5B5D61;
  font-size: 0.8125rem;
  line-height: 1.25;
  max-width: 100%;
  line-height: 0.875rem;
}

<div
  aria-selected={false}
  className="emotion-11 emotion-12"
  fontSize="size4"
  id="menu-item_clearance_state_ready_to_clear"
  label="Ready to Clear"
  name="clearance_state"
  onClick={[Function]}
  onKeyPress={[Function]}
  onSelect={[Function]}
  role="option"
  value="ready_to_clear"
>
  <input
    className="emotion-0"
    id="menu-item_clearance_state_ready_to_clear"
    name="clearance_state"
    type="hidden"
    value="ready_to_clear"
  />
  <div
    className="emotion-9 emotion-4"
    width="100%"
  >
    <div
      className="emotion-3 emotion-4"
    >
      <span
        className="emotion-1 emotion-2"
        color="inherit"
        fontSize="size4"
        overflow="hidden"
      >
        Ready to Clear
      </span>
    </div>
    <div
      className="emotion-7 emotion-4"
      overflow="hidden"
    >
      <span
        className="emotion-5 emotion-2"
        color="text.muted"
        fontSize="size3"
      >
        mpd@catchandrelease.com
      </span>
    </div>
  </div>
</div>
`;

exports[`<ControlledMenuItem /> renders a selected MenuItem properly 1`] = `
.emotion-11 {
  font-family: Motiva Sans;
  font-size: 0.875rem;
  padding: 0.75rem;
  box-sizing: border-box;
  background: #FFFFFF;
  color: #25272D;
  cursor: pointer;
}

.emotion-11:hover,
.emotion-11:focus {
  background: #F8F9F9;
}

.emotion-0 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  -webkit-clip-path: inset(50%);
  clip-path: inset(50%);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  white-space: nowrap;
  width: 1px;
}

.emotion-9 {
  font-family: Motiva Sans;
  box-sizing: border-box;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.emotion-3 {
  font-family: Motiva Sans;
  max-width: calc(100% - 120px - 8px);
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
}

.emotion-1 {
  font-family: Motiva Sans;
  color: inherit;
  font-size: 0.875rem;
  line-height: 1.25;
  min-width: 120px;
  overflow: hidden;
}

.emotion-7 {
  font-family: Motiva Sans;
  overflow: hidden;
  padding-left: 8px;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0 1 auto;
  -ms-flex: 0 1 auto;
  flex: 0 1 auto;
}

.emotion-5 {
  font-family: Motiva Sans;
  color: #5B5D61;
  font-size: 0.8125rem;
  line-height: 1.25;
  max-width: 100%;
  line-height: 0.875rem;
}

<div
  aria-selected={false}
  className="emotion-11 emotion-12"
  fontSize="size4"
  id="menu-item_clearance_state_ready_to_clear"
  label="Ready to Clear"
  name="clearance_state"
  onClick={[Function]}
  onKeyPress={[Function]}
  onSelect={[Function]}
  role="option"
  selected={true}
  value="ready_to_clear"
>
  <input
    className="emotion-0"
    id="menu-item_clearance_state_ready_to_clear"
    name="clearance_state"
    type="hidden"
    value="ready_to_clear"
  />
  <div
    className="emotion-9 emotion-4"
    width="100%"
  >
    <div
      className="emotion-3 emotion-4"
    >
      <span
        className="emotion-1 emotion-2"
        color="inherit"
        fontSize="size4"
        overflow="hidden"
      >
        Ready to Clear
      </span>
    </div>
    <div
      className="emotion-7 emotion-4"
      overflow="hidden"
    >
      <span
        className="emotion-5 emotion-2"
        color="text.muted"
        fontSize="size3"
      >
        mpd@catchandrelease.com
      </span>
    </div>
  </div>
</div>
`;
